### DISCUSSION TITLE
Making ThumbmarkJS work in Next.js

### DISCUSSION THREAD

**Discussion by ilkkapeltola** [thumbmarkjs/thumbmarkjs#5] [Ideas]
ThumbmarkJS relies on the `navigator` for some information.
But this isn't available in Next.js.

What's the general approach to a situation like this?
Or how would ThumbmarkJS go around this issue?
Can this work to begin with?

`navigator` is used in many places in this library to check certain things. It exposes the userAgent for example among many useful fingerprinting sources.

Ideas?

**Comment by ilkkapeltola** [comment]
@noe how does your setup work with Next.js?
Are you nevertheless planning to run the library in a browser environment eventually? And the dynamic importing is so the calls to navigator won't be done when it isn't there?

A pretty simple approach would be to stop the calls to the component functions when navigator isn't present. That way at least the library could be imported, but it would never reach a call to navigator. Or something a little more sophisticated.

Maybe I'm off the mark. 😀

**Reply by noe** [reply]
Next.js is a framework built on top of React that allows you to do both server-side and client-side rendering. In general, everything is rendered server-side, but you can use client-side-specific stuff, e.g. React's `useEffect`.

In general, when using Next.js, one you be aware if some piece of code is meant to be run only on the client side. In the case of ThumbmarkJS it is natural that it can only run on the client side, so no problem with it using `navigator`. However, it seems `ThumbmarkJS` uses `navigator` when it is imported. This is a problem in Next.js because that import may happen server-side, where `navigator` is not available.

The specific problem seems to be in [`fonts.ts`](https://github.com/ilkkapeltola/thumbmarkjs/blob/main/src/components/fonts/fonts.ts#L144), where you check if the browser is Firefox in the body of the typescript file:

```ts
if (getBrowser().name != 'Firefox')
    includeComponent('fonts', getFontMetrics)
```

`getBrowser` uses `navigator`, therefore the error when run server-side.

This could probably be fixed by adding an extra check before, to avoid running `getBrowser` server-side:

```ts
if (typeof window !== 'undefined' && getBrowser().name != 'Firefox')
    includeComponent('fonts', getFontMetrics)
```

The extra check is a standard approach to see if the code is running server-side.

I have checked if the same check is used elsewhere but found no other instance of it. I may have missed it, though.

**Reply by noe** [reply]
I forgot to mention that ThumbmarkJS does work as-is in Next.js when used with a dynamic import, like this:

```ts
const ThumbmarkJS = import('thumbmarkjs');
const {getFingerprint} = await ThumbmarkJS;
const fingerprint = await getFingerprint();
```

**Reply by ilkkapeltola** [reply]
Right of course. This makes sense. And that is indeed currently the only place where navigator is accessed already on import.

This is really good feedback. I hadn't considered this use case. I suppose this is bad error handling inside the getBrowser().
I've modified the getBrowser() function to return 'unknown' when navigator isn't present. This should do the same as with checking for if window is undefined.

But something to think about for me, how I ensure that the package stays safe to import.

the version 0.11.4 should now have this particular case fixed.

**Comment by deadcoder0904** [comment]
> ⨯ node_modules\.pnpm\@thumbmarkjs+thumbmarkjs@0.13.2\node_modules\@thumbmarkjs\thumbmarkjs\dist\components\webgl\webgl.js (42:13) @ document
 ⨯ unhandledRejection: ReferenceError: document is not defined

i get this error when trying with next.js on the latest version.

i think the library assumes `document` exists without a check:

see https://www.bugpilot.com/guides/en/fixing-the-document-is-not-defined-error-in-next-js-98b7 for more info.

it did work last week when i tried. so i get [this commit](https://github.com/thumbmarkjs/thumbmarkjs/commit/c473efdfa7bee50f13973ffec47699f2597a440a) is the issue? can you add a check to `document`? maybe need a linter rule to not use document directly without a check to not have errors in frameworks like next.js.

dynamic import doesn't work for me like the above solution or this:

```ts
import dynamic from "next/dynamic";

const thumbmarkjs = dynamic(
  () => {
    return import("@thumbmarkjs/thumbmarkjs")
  },
  { ssr: false }
);
```

**edit:** alright, this worked.

```ts
"use client"

import React from "react"
import { hasWindow } from "std-env"

import { useStore } from "@/app/store/zustand"

export function Fingerprint() {
	const { setFingerprint } = useStore()

	React.useEffect(() => {
		async function fn() {
			if (hasWindow) {
				const { getFingerprint } = await import("@thumbmarkjs/thumbmarkjs")
				const visitorId = await getFingerprint()
				setFingerprint(visitorId)
			}
		}

		fn()
	}, [setFingerprint])
	return null
}
```

**Reply by ilkkapeltola** [reply]
Oh no, yeah okay, I think you're right. I have to re-think my approach I guess...

I'll implement a check for document to exist before creating those elements.

**Reply by ilkkapeltola** [reply]
by the way, opening an issue is clearer.
This discussion was already closed, so I missed the message.

**Reply by ilkkapeltola** [reply]
@deadcoder0904 this is now fixed :)

**Reply by deadcoder0904** [reply]
@ilkkapeltola can you show the code in the README or here?

**Reply by ilkkapeltola** [reply]
You mean provide an example of how to implement this in a Next.js project?

**Reply by deadcoder0904** [reply]
@ilkkapeltola ignore it. i thought the code would change but i guess it would be the same now even if its next.js. highly recommend though setting a top-level `if` on `document` otherwise future releases might break next.js again. like [this](https://env.t3.gg/docs/nextjs) works for `.env`, something similar.

**Reply by ilkkapeltola** [reply]
I would like to keep the module importable outside of useEffect. [this works](https://github.com/thumbmarkjs/thumbmarkjs/blob/main/examples/nextjs.tsx) and is IMO cleaner. Not really necessary, but still. The `getFingerprint()` function should though return a useful error when it is attempted to run server side.

What I should do, is integration testing. But I'm not sure what makes most sense in that. I created now a simple Next.js project that includes this library for testing purposes, but ideally, I would run a new version always against a test environment before publishing.

**Reply by deadcoder0904** [reply]
@ilkkapeltola i think next.js broke again.

```bash
142.7   There are types at '/app/node_modules/@thumbmarkjs/thumbmarkjs/dist/types/index.d.ts', but this result could not be resolved when respecting package.json "exports". The '@thumbmarkjs/thumbmarkjs' library may need to update its package.json or typings.
142.7
142.7   12 |            async function fn() {
142.7   13 |                    if (hasWindow) {
142.7 > 14 |                            const { getFingerprint } = await import("@thumbmarkjs/thumbmarkjs")
142.7      |                                                                    ^
142.7   15 |                            const visitorId = await getFingerprint()
142.7   16 |                            setFingerprint(visitorId)
142.7   17 |                    }
```

`v0.13.5` works but after that, it doesn't work. tried all versions until `v0.14.1`. you definitely need to test it to solve this once & for all.

**Comment by ilkkapeltola** [comment]
I think this must be something to do with the way I've set up Rollup. I'll
see if I can fix this in the evening. I switched from webpack to rollup and
in that, ended up changing also the way the NPM package is packaged. I
suspect that's the issue.

ma 4. maalisk. 2024 klo 16.23 akshay kadam (a2k) ***@***.***)
kirjoitti:

**Reply by ilkkapeltola** [reply]
By the way, I don't think this will make the library fail, just the typings aren't working. It should still run right.

**Reply by ilkkapeltola** [reply]
And, also, I did test this locally with a next.js app without any issues. Not sure why I didn't capture it.
